examples/main/main.cpp

```diff
@@ -363,14 +338,42 @@ int main(int argc, char ** argv) {
                 n_past += n_eval;
             }
 
-            if (embd.size() > 0 && !path_session.empty()) {
-                session_tokens.insert(session_tokens.end(), embd.begin(), embd.end());
-                n_session_consumed = session_tokens.size();
+            // save session after context swap
+            if (!path_session.empty() && needs_swap) {
+                int n_tokens = n_past - params.n_keep;
+                if (!llama_append_session_file(
+                        ctx, path_session.c_str(), params.n_keep,
+                        last_n_tokens.data() + last_n_tokens.size() - n_tokens, n_tokens)) {
+                    fprintf(stderr, "%s: error: unable to write to session file '%s'\n",
+                        __func__, path_session.c_str());
+                    return 1;
+                }
+
+                n_session_write_past = n_past;
```

Comment:
And whole prompt should be >= size of n_ctx to save session, but using whole prompt next time will give us error "prompt is too long"

Reply:
Yeah, per my other comment I'll probably drop this. To be clear though, if you're talking about what's saved in the session, the state changes are incremental. The original prompt state will always be there in the first segment of the session regardless of what comes after and that's all the startup code loads.